### PR TITLE
feat: faas plugin refactoring with url path forwarding

### PR DESCRIPTION
**Pull request by bisakhmondal**
### What this PR does / why we need it:

This PR is based on the idea of creating a single generic upstream where each faas plugin implements an authorization schema and perform the header processing.

GOAL: Reduce the same code over multiple plugins (mainly faas)

**How developers should write new (faas) plugins:**
1. They should define plugin-specific authorization schema & metadata schema(if required)
2. Write a `request_processor` function on the plugin lua file that take `plugin conf, ngx context and the URL params (that contains headers, body, urlpath, query params etc. etc.)". The function can perform a series of operations including update/adding the necessary authorization headers, performing encoding/decoding (base64 or anything) on the body.
3. And yes, that's it. (see the azure-functions Lua code for example)

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: apisix/plugins/serverless/generic-upstream.lua [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        end
+
+        -- forward the url path came through the matched uri
+        local url_decoded = url.parse(conf.function_uri)
+        local path = url_decoded.path or "/"
+
+        if ctx.curr_req_matched and ctx.curr_req_matched["*"] then
+            local end_path = ctx.curr_req_matched["*"]
+
+            if path:sub(-1, -1) == "/" or end_path:sub(1, 1) == "/" then
```
Comment (spacewander) [inline apisix/plugins/serverless/generic-upstream.lua:78]: Using `:byte` would be better.
BTW, it is not a good idea to refactor & add feature in the same PR.

File: apisix/plugins/serverless/generic-upstream.lua [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        if err then
+            core.log.error("error while reading request body: ", err)
+            return 400
+        end
+
+        -- forward the url path came through the matched uri
+        local url_decoded = url.parse(conf.function_uri)
+        local path = url_decoded.path or "/"
+
+        if ctx.curr_req_matched and ctx.curr_req_matched["*"] then
```
Comment (spacewander) [inline apisix/plugins/serverless/generic-upstream.lua:75]: We should use `:ext` instead of `*`

File: t/plugin/azure-functions.t [hunk 3]
```diff
@@ -375,3 +381,43 @@ ngx.say("Authz-Header - " .. headers["x-functions-key"] or "")
...
+            -- creating a semi path route
+            local code, body = t('/apisix/admin/routes/1',
+                 ngx.HTTP_PUT,
+                 [[{
+                        "plugins": {
+                            "azure-functions": {
+                                "function_uri": "http://localhost:8765/api"
+                            }
+                        },
+                        "uri": "/azure/**"
```
Comment (spacewander) [inline t/plugin/azure-functions.t:401]:
```suggestion
                        "uri": "/azure/*"
```

File: apisix/plugins/serverless/generic-upstream.lua (hunk 1 above)
Comment (bisakhmondal) [reply apisix/plugins/serverless/generic-upstream.lua:78]: for byte you mean like this 
`if path:byte(-1, -1) == string.byte"/" or end_path:byte(1, 1) == string.byte"/" then`

The PRs are interdependent and require one PR to be merged before another. Do you think I should split it up into 2 PRs?

Comment (spacewander) [reply apisix/plugins/serverless/generic-upstream.lua:78]: Like `if path:byte(-1) == string.byte("/") or end_path:byte(1) == string.byte("/") then`.

We can accept this one now.
Next time, please don't mix up refactor & new feature.

Comment (bisakhmondal) [reply apisix/plugins/serverless/generic-upstream.lua:78]: Cool! Thanks boss

File: t/plugin/azure-functions.t [hunk 4]
```diff
@@ -375,3 +381,43 @@ ngx.say("Authz-Header - " .. headers["x-functions-key"] or "")
...
+            )
+            if code >= 300 then
+                ngx.status = code
+                ngx.say("fail")
+                return
+            end
+
+            ngx.say(body)
+
+            local code, _, body = t("/azure/httptrigger", "GET")
```
Comment (spacewander) [inline t/plugin/azure-functions.t:412]: Let's add another test which contains multiple levels of paths, like `/azure/http/trigger`.

File: apisix/plugins/serverless/generic-upstream.lua (hunk 2 above)
Comment (bisakhmondal) [reply apisix/plugins/serverless/generic-upstream.lua]: Agree. I was using `**` that's why the extra arg was received through curr_req_matched. Updated

File: t/plugin/azure-functions.t (hunk 3 above)
Comment (bisakhmondal) [reply t/plugin/azure-functions.t]: Done

File: t/plugin/azure-functions.t (hunk 4 above)
Comment (bisakhmondal) [reply t/plugin/azure-functions.t:424]: Done.

Comment (bisakhmondal) [reply t/plugin/azure-functions.t:424]: As discussed offline the following cases have been covered
multiple levels of paths
multiple slashes in the paths
the wildcard matches an empty part, like /azure

File: apisix/plugins/serverless/generic-upstream.lua (hunk 1 above)
Comment (bisakhmondal) [reply apisix/plugins/serverless/generic-upstream.lua]: Done

File: t/plugin/azure-functions.t [hunk 5]
```diff
@@ -375,3 +393,100 @@ ngx.say("Authz-Header - " .. headers["x-functions-key"] or "")
...
+            ngx.print(body)
+        }
+    }
+--- response_body
+passed
+invocation /api/httptrigger successful
+
+
+
+=== TEST 11: check multilevel url path forwarding
```
Comment (spacewander) [inline t/plugin/azure-functions.t:439]: Next time, we can use a table drive test for those cases. See https://github.com/apache/apisix/blob/7242216076f1df4225ab5d2d742ed7fb23383306/t/plugin/gzip.t#L498

**Review by spacewander — APPROVED** [review]

**Review by tzssangglass — APPROVED** [review]

**Pull request merged**